### PR TITLE
Build: dump user's command into a shell script before running it

### LINKED ISSUE #10103: Build: `build.jobs.post_build` produces mysterious shell syntax errors
## Details

* Read the Docs project URL: https://readthedocs.org/projects/mmref/
* Build URL (if applicable): https://readthedocs.org/projects/mmref/builds/19671469/ and https://readthedocs.org/projects/mmref/builds/19671433/ and most previous  builds
* Read the Docs username (if applicable): https://readthedocs.org/profiles/rptb1/

## Expected Result

A post_build script `if test -n "A"; then echo "B"; fi` should not produce a syntax error.
A post_build script with `sh -c "foo"` in it should run the shell with the argument.

## Actual Result

`/bin/sh: 1: Syntax error: "then" unexpected`
`sh: 0: -c requires an argument`

I'm following the manual at ["Extend the build process"](https://docs.readthedocs.io/en/stable/build-customization.html#extend-the-build-process) and trying to insert a simple shell script at the post_build step.

A script that works locally (Ubuntu 22) is somehow mangled and causes syntax errors when inserted into the .readthedocs.yml at the post_build stage.  I have tried many ways of writing the script and quoting it in YAML, to try to eliminate YAML as the culprit:

1. https://github.com/Ravenbrook/mps/pull/187/commits/4146715a1253a9ba7dcb8e528997545d52bb7ab2
2. https://github.com/Ravenbrook/mps/pull/187/commits/9fbc6224dbe1b645568fd32a938c77f84ef67d3a
3. https://github.com/Ravenbrook/mps/pull/187/commits/2a642594a525a3cbcd0d9c10f686bf6b6e43307c
4. https://github.com/Ravenbrook/mps/pull/187/commits/7fdbbd3ff01544484ba5a518b46c96dc9c98f1f3
5. https://github.com/Ravenbrook/mps/pull/187/commits/82f0d180a1bb0a7909d94eee22ab4a7b71a7dea3
6. https://github.com/Ravenbrook/mps/pull/187/commits/044e790f24217ad709dbc8ced9c35febaf02f0e9
7. https://github.com/Ravenbrook/mps/pull/187/commits/4a07b8418a4e5b1036dd50f13552098308972fb2
8. https://github.com/Ravenbrook/mps/pull/187/commits/1106c9cd066c9638ccf2f7772083384bb526673f

The [last one](https://github.com/Ravenbrook/mps/pull/187/commits/1106c9cd066c9638ccf2f7772083384bb526673f) is especially mysterious.

Is there some sort of mangling of the script going on, or am I just going blind?

None of these scripts produce syntax errors when I feed them to /bin/sh on my local Ubuntu.  I have written similar scripts for multiple CI systems without coming across this problem before.

Thanks.

### PR DESCRIPTION
**Pull request by humitos**
We attemped to fix this issue multiple times and we found problems when deploying those changes to production.

This not perfect, but simple solution, could help us to support multi-line commands for now without manipulating "too much" the `PATH` variable, that was the one causing us issues.

Let me know what you think about this approach and how to move forward if you prefer a different one.

Closes #10103
Reference #10206 
Reference #10172 
Reference #10133

### CODE REVIEW

File: readthedocs/doc_builder/environments.py [hunk 1]
```diff
@@ -432,6 +372,44 @@ def _escape_command(self, cmd):
...
+        script_content = f"""
+#!/bin/bash
+
+export PATH={bin_path}$PATH
+{command}
+"""
+        tarstream = io.BytesIO()
+        tar = tarfile.TarFile(fileobj=tarstream, mode="w")
+        script_data = script_content.encode("utf8")
+        tarinfo = tarfile.TarInfo(name="readthedocs.sh")
```
Comment (ericholscher) [inline readthedocs/doc_builder/environments.py:401]: I think we probably want to use a random name for this file, so we don't end up with some kind of weird docker error running old commands?

File: readthedocs/doc_builder/environments.py [hunk 2]
```diff
@@ -432,6 +372,44 @@ def _escape_command(self, cmd):
...
+        tarinfo.size = len(script_data)
+        tarinfo.mtime = time.time()
+        tar.addfile(tarinfo, io.BytesIO(script_data))
+        tar.close()
+        tarstream.seek(0)
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/doc_builder/environments.py:406]: Unfortunate, the DockerAPI requires this to be a `tarfile` 🤷🏼 

https://docker-py.readthedocs.io/en/stable/containers.html?highlight=put_archive#docker.models.containers.Container.put_archive

**Comment by humitos** [comment]
I'm not sure if it makes sense to move forward with this PR or not due the doubts and concerns of this implementation. However, I ran out of ideas trying to solve this problem. If there are better ideas, I'm happy to implement them.

**Comment by bensze01** [comment]
How about just passing the multiline script as a parameter to `sh -c` instead of writing it out to a file? eg.
```bash
sh -c 'export PATH={bin_path}$PATH
{command}'
```
Seems like it would require fewer changes.

**Pull request closed without merging**